### PR TITLE
Validation error tagging 4

### PR DESCRIPTION
**Pull request by zdrve**
Here for previewing

Still to do:

- merge the previous branch to master
- get zendesk-kennel using it
- remove the four references to `validate` in OptionalValidations
- run zendesk-kennel against _this_ branch to check that the migration isn't going to be too hard
- fill in test coverage gaps (currently 18 lines)

### CODE REVIEW

File: lib/kennel/models/dashboard.rb [hunk 1]
```diff
@@ -230,7 +230,9 @@ def validate_json(data)
 
         # Avoid diff from datadog presets sorting.
         presets = data[:template_variable_presets]
-        invalid! "template_variable_presets must be sorted by name" if presets && presets != presets.sort_by { |p| p[:name] }
+        if presets && presets != presets.sort_by { |p| p[:name] }
+          invalid! :template_variable_presets_must_be_sorted, "template_variable_presets must be sorted by name"
```
Comment (grosser) [inline lib/kennel/models/dashboard.rb:234]:
```suggestion
          invalid! :template_variable_presets_unsorted, "template_variable_presets must be sorted by name"
```

File: lib/kennel/models/monitor.rb [hunk 2]
```diff
@@ -205,26 +205,26 @@ def validate_json(data)
         super
 
         if data[:name]&.start_with?(" ")
-          invalid! "name cannot start with a space"
+          invalid! :name_must_not_start_with_space, "name cannot start with a space"
```
Comment (grosser) [inline lib/kennel/models/monitor.rb:208]:
```suggestion
          invalid! :name_starts_with_space, "name cannot start with a space"
```

File: lib/kennel/models/record.rb [hunk 3]
```diff
@@ -211,8 +211,12 @@ def resolve_link(sought_tracking_id, sought_type, id_map, force:)
         end
       end
 
-      def invalid!(message)
-        unfiltered_validation_errors << ValidationMessage.new(message)
+      def invalid!(tag, message = nil)
+        if message.nil?
+          message = tag
+          tag = OptionalValidations::NO_TAG_CHOSEN
```
Comment (grosser) [inline lib/kennel/models/record.rb:217]:
```suggestion
          tag = OptionalValidations::NO_TAG_CHOSEN # TODO: fail here and clean up all usages
```

File: lib/kennel/optional_validations.rb [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
+
+    UnignorableTag = Struct.new(:inspect) # rubocop:disable Lint/StructNewOverride
+    NO_TAG_CHOSEN = UnignorableTag.new("<no tag chosen>")
+    NO_TAG = UnignorableTag.new("<not ignorable>")
 
     def self.included(base)
-      base.settings :validate
-      base.defaults(validate: -> { true })
+      base.settings :ignored_errors, :validate
+      base.defaults(ignored_errors: -> { [] }, validate: true)
```
Comment (grosser) [inline lib/kennel/optional_validations.rb:12]: should we call this `ignored_validations` ?

File: lib/kennel/optional_validations.rb [hunk 5]
```diff
@@ -32,46 +48,39 @@ def validate_json(data)
...
-          unfiltered_validation_errors
-        else
-          if mode == "show"
-            unfiltered_validation_errors.each do |err|
-              Kennel.out.puts "#{safe_tracking_id} `validate: false` suppressing error: #{err.text.gsub("\n", " ")}"
+        to_report =
+          if ENV["NO_IGNORED_ERRORS"]
+            # Turn off all suppressions, to see what errors are actually being suppressed
+            unfiltered_validation_errors
+          elsif !validate
```
Comment (grosser) [inline lib/kennel/optional_validations.rb:70]:
```suggestion
          elsif !validate # TODO: all these usages should be converted to named ignores
```

**Review by grosser — COMMENTED** [review]
shape looks about right
would love some consistent naming
either "name says what is wrong" or "name says what needs to change"

File: lib/kennel/optional_validations.rb (hunk 5 above)
Comment (zdrve) [reply lib/kennel/optional_validations.rb:70]: This is going away before merge

File: lib/kennel/optional_validations.rb (hunk 4 above)
Comment (zdrve) [reply lib/kennel/optional_validations.rb:12]: If you like. That's a good name.

Comment (zdrve) [reply lib/kennel/optional_validations.rb:12]: Welll.... we don't actually ignore the validation, we still run the validation and then ignore the error.

But it's not something I have a strong opinion on. Some combination of "ignore/skip/expect/suppress"\[ed\] "errors/validations/validation_errors" works.

File: lib/kennel/optional_validations.rb (hunk 5 above)
Comment (zdrve) [reply lib/kennel/optional_validations.rb:70]: This is to be removed before merge.

**Pull request closed without merging**